File: HRMS/HRMS.Web/Modules/HR/EmployeeSalary/EmployeeSalaryForm.cs
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;

namespace HRMS.HR.Forms;

[FormScript("HR.EmployeeSalary")]
[BasedOnRow(typeof(EmployeeSalaryRow), CheckNames = true)]
public class EmployeeSalaryForm
{
    public int EmployeeId { get; set; }
    public int GradeId { get; set; }
    public DateTime EffectiveDate { get; set; }
    public decimal BasicSalary { get; set; }
    [ReadOnly(true)]
    public decimal GrossSalary { get; set; }
    
    [Category("Salary Components")]
    [EditorType("HRMS.HR.EmployeeSalaryDetailsEditor")]
    public List<EmployeeSalaryDetailsRow> DetailList { get; set; }
    
    public bool IsActive { get; set; }
}

File: HRMS/HRMS.Web/Modules/ServerTypes/HR/EmployeeSalaryForm.ts
import { BooleanEditor, DateEditor, DecimalEditor, initFormType, LookupEditor, PrefixedContext, Widget } from "@serenity-is/corelib";

export interface EmployeeSalaryForm {
    EmployeeId: LookupEditor;
    GradeId: LookupEditor;
    EffectiveDate: DateEditor;
    BasicSalary: DecimalEditor;
    GrossSalary: DecimalEditor;
    DetailList: Widget;
    IsActive: BooleanEditor;
}

export class EmployeeSalaryForm extends PrefixedContext {
    static readonly formKey = 'HR.EmployeeSalary';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!EmployeeSalaryForm.init) {
            EmployeeSalaryForm.init = true;

            var w0 = LookupEditor;
            var w1 = DateEditor;
            var w2 = DecimalEditor;
            var w3 = Widget;
            var w4 = BooleanEditor;

            initFormType(EmployeeSalaryForm, [
                'EmployeeId', w0,
                'GradeId', w0,
                'EffectiveDate', w1,
                'BasicSalary', w2,
                'GrossSalary', w2,
                'DetailList', w3,
                'IsActive', w4
            ]);
        }
    }
}